src/VectorBundle.jl

```diff
@@ -0,0 +1,221 @@
+
+"""
+    VectorSpaceType
+
+Abstract type for tangent space, cotangent space, their tensor products,
+exterior products etc.
+"""
+abstract type VectorSpaceType end
+
+struct TangentSpaceType <: VectorSpaceType end
+struct CotangentSpaceType <: VectorSpaceType end
+
+TCoTSpaceType = Union{TangentSpaceType, CotangentSpaceType}
+
+const TangentSpace = TangentSpaceType()
+const CotangentSpace = CotangentSpaceType()
+
+"""
+    similar_result_type(M::Manifold, VS::VectorSpaceType, f, args::NTuple{N,Any}) where N
+
+Returns type of element of the array that will represent the result of
+function `f` for representing an operation with result in the vector space `VS`
+for manifold `M` on given arguments (passed at a tuple).
+"""
+function similar_result_type(M::Manifold, VS::VectorSpaceType, f, args::NTuple{N,Any}) where N
+    T = typeof(reduce(+, one(eltype(eti)) for eti ∈ args))
+    return T
+end
+
+"""
+    similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+
+Allocates an array for the result of function `f` that is
+an element of the vector space of type `VS` on manifold `M`
+and arguments `x...` for implementing the non-modifying operation
+using the modifying operation.
+"""
+function similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+    T = similar_result_type(M, VS, f, x)
+    return similar(x[1], T)
+end
+
+
+"""
+    manifold_dimension(M::Manifold, VS::VectorSpaceType)
+
+Dimension of the vector space of type `VS` over manifold `M`.
+"""
+function manifold_dimension(M::Manifold, VS::VectorSpaceType)
+    error("manifold_dimension not implemented for manifold $(typeof(M)) and vector space of type $(typeof(VS)).")
+end
+
+manifold_dimension(M::Manifold, ::TCoTSpaceType) = manifold_dimension(M)
+
+function representation_size(M::Manifold, VS::TCoTSpaceType)
```

Comment:
I think the point of these functions is to provide much of the same functionality for vector spaces as we provide for manifolds. I guess I'm just concerned about all of these different signatures. Wouldn't it be cleaner to define vector space as a `Manifold` typed on and containing the manifold `M`? I doubt that'll add much, if any, overhead, and it's more consistent. Then all of these functions could maintain the same signature as the rest of the interface.

Reply:
I wrote about this above: https://github.com/JuliaNLSolvers/Manifolds.jl/pull/31#discussion_r302963120 so maybe we can discuss it there.